### PR TITLE
Extend config options

### PR DESCRIPTION
**Pull request by ppacher**
Extends the config option system with new properties. Also adds a bit more type safety.

### CODE REVIEW

File: config/option.go [hunk 1]
```diff
@@ -34,32 +37,174 @@ func getTypeName(t uint8) string {
...
+}
+
+// GetAnnotationUnsafe is like GetAnnotation but does not lock the option.
+func (option *Option) GetAnnotationUnsafe(key string) (interface{}, bool) {
+	if option.Annotations == nil {
+		return nil, false
+	}
+	val, ok := option.Annotations[key]
+	return val, ok
+}
```
Comment (dhaavi) [inline config/option.go:206]: Is there a reason for exporting the `Unsafe` method variants?
I'd like to follow a convention where exported functions are safe to use in regards to locking, but internal functions usually aren't.

File: config/option.go [hunk 2]
```diff
@@ -34,32 +37,174 @@ func getTypeName(t uint8) string {
...
+type Annotations map[string]interface{}
+
+// Well known annotations defined by this package.
+const (
+	// DisplayHintAnnotation provides a hint for the user
+	// interface on how to render an option.
+	// The value of DisplayHintAnnotation is expected to
+	// be a string. See DisplayHintXXXX constants below
+	// for a list of well-known display hint annotations.
+	DisplayHintAnnotation = "sfng/portbase:ui:display-hint"
```
Comment (dhaavi) [inline config/option.go:70]: I would suggest using `safing/portbase` as that is part of our Go project import path.
Go itself also uses the domain for identifying a package, not sure if we need this level of uniqueness, probably not.

File: config/validate.go [hunk 3]
```diff
@@ -28,7 +29,42 @@ func (vc *valueCache) getData(opt *Option) interface{} {
...
+		// loading int's from the configuration JSON does not perserve the correct type
+		// as we get float64 instead. Make sure to convert them before.
+		if reflect.TypeOf(val.Value).ConvertibleTo(valueType) {
+			compareAgainst = reflect.ValueOf(val.Value).Convert(valueType).Interface()
+		}
+		if compareAgainst == value {
+			return nil
+		}
+
+		//
```
Comment (dhaavi) [inline config/validate.go:50]: Forgot to add something here?

File: config/validate.go [hunk 4]
```diff
@@ -28,7 +29,42 @@ func (vc *valueCache) getData(opt *Option) interface{} {
...
+		compareAgainst := val.Value
+		valueType := reflect.TypeOf(value)
+
+		// loading int's from the configuration JSON does not perserve the correct type
+		// as we get float64 instead. Make sure to convert them before.
+		if reflect.TypeOf(val.Value).ConvertibleTo(valueType) {
+			compareAgainst = reflect.ValueOf(val.Value).Convert(valueType).Interface()
+		}
+		if compareAgainst == value {
+			return nil
```
Comment (dhaavi) [inline config/validate.go:47]: Return on first success?

File: config/validate.go [hunk 5]
```diff
@@ -28,7 +29,42 @@ func (vc *valueCache) getData(opt *Option) interface{} {
...
+		}
+
+		//
+		if reflect.DeepEqual(val.Value, value) {
+			return nil
```
Comment (dhaavi) [inline config/validate.go:52]: Return on first success?

File: modules/subsystems/module.go [hunk 6]
```diff
@@ -25,10 +26,22 @@ var (
...
 	modules.EnableModuleManagement(handleModuleChanges)
 
 	// register module and enable it for starting
+	// TODO(ppacher): we have a very problematic dependecy to base here:
+	// 				  the "base" module is actually part of portmaster
+	//				  which in turn depends on portbase. This works as there's
+	//				  no import cycle but it basically makes the subsystems package
+	//                usable without bringing in a dependency on the complete portmaster!
+	//			      Consider getting rid of the "base" dependency or moving the whole
+	//				  subsystem layer to portmaster.
```
Comment (dhaavi) [inline modules/subsystems/module.go:44]: This is actually intended, as everyone who starts using portbase should bring their own base module. The base module is primarily used for setup procedures that have to happen before anything else. If that is not needed, a developer can supply a no-op base module. I admit, this is not documented anywhere, afaik.

**Review by dhaavi — CHANGES_REQUESTED** [review]
Looks good so far, had some questions.

File: config/validate.go [hunk 7]
```diff
@@ -28,7 +29,42 @@ func (vc *valueCache) getData(opt *Option) interface{} {
 	}
 }
 
+func validatePossibleValues(opt *Option, value interface{}) error {
```
Comment (dhaavi) [inline config/validate.go:32]: Please add comment.

**Review by dhaavi — CHANGES_REQUESTED** [review]

File: config/option.go [hunk 8]
```diff
@@ -34,32 +37,184 @@ func getTypeName(t uint8) string {
...
+	PossibleValues []PossibleValue `json:",omitempty"`
+	// Annotations adds additional annotations to the configuration options.
+	// See documentation of Annotations for more information.
+	// Annotations is considered mutable and setting/reading annotation keys
+	// must be performed while the option is locked.
+	Annotations Annotations
 
 	activeValue         *valueCache // runtime value (loaded from config file or set by user)
-	activeDefaultValue  *valueCache // runtime default value (may be set internally)
+	activeDefaultValue  *valueCache // e default value (may be set internally)
```
Comment (dhaavi) [inline config/option.go:174]: Typo?

File: config/registry.go [hunk 9]
```diff
@@ -11,6 +12,26 @@ var (
...
+	optionsLock.Lock()
+	defer optionsLock.Unlock()
+
+	for _, opt := range options {
+		opt.Lock()
+		if err := fn(opt); err != nil {
+			opt.Unlock()
+			return err
+		}
+		opt.Unlock()
```
Comment (dhaavi) [inline config/registry.go:30]: `ForEachOption does NOT lock opt when calling fn.` 🤔

**Review by dhaavi — CHANGES_REQUESTED** [review]

File: config/registry.go (hunk 9 above)
Comment (ppacher) [reply config/registry.go:30]: ^^ that was a merge conflict and I've overseen the locking part :D Will fix it

File: config/option.go (hunk 8 above)
Comment (ppacher) [reply config/option.go:174]: yes, I guess so :facepalm:

**Review by dhaavi — APPROVED** [review]

**Pull request merged**